design-proposals/feature-lifecycle.md

```diff
@@ -0,0 +1,327 @@
+# KubeVirt Feature Lifecycle
+
+## Summary
+
+KubeVirt requires a clear policy on how features are introduced,
+evaluated and finally graduated or removed.
+
+This proposal defines the steps and policies to follow in order
+to manage a feature and its lifecycle in KubeVirt.
+
+The proposal is focusing on introducing features in
+a stable API (CRD) version, e.g. `kubevirt.io/v1`.
+
+## Overview
+
+### Motivation
+KubeVirt has grown into a matured virtualization management solution
+with a large set of features.
+
+New features are being proposed and added regularly to its portfolio.
+
+With time, the challenge of supporting and maintaining such a large
+set of features raised the need to re-examine their relevance.
+It also raised the need to examine with more care features graduation.
+
+The KubeVirt community has tried to control the flow of features
+informally through feature-gates, similar to Kubernetes.
+However, as time passed, several challenges presented themselves:
+- Evaluated features rarely got graduated to GA or removed, causing
+  feature consumption to be risky for users and a maintenance burden
+  for the project contributors.
+  > **Note**: As of this writing (pre v1.2), there are 37 FGs,
+  > out of which 5 GA-ed and 2 marked for deprecation.
+  > For more information, explore the
+  > [source](https://github.com/kubevirt/kubevirt/blob/5ff12ae931cefd81514ec96f97a189ab2c179ad7/pkg/virt-config/feature-gates.go).
+- We do not have agreed-upon procedures on how and when features
+  graduate or discontinue.
+  This causes each feature to take different approaches, possibly
+  surprising users.
+
+We conclude that the current use of FGs is insufficient
+due to the lack of well-defined processes and policy on how a feature
+should progress in its lifecycle.
+
+Eventually we would like to see features being evaluated carefully
+before they are introduced, while they are experimented with and
+proven to be actually in use (and useful) before graduating.
```

Comment:
It's not always easy to confirm whether a feature is actively being used. Furthermore, certain features, like recovery tools, might not see frequent use, but they can be crucial in critical situations.

Reply:
Okay, I understand this point. The reason I raised this is because, as mentioned [here](https://github.com/kubevirt/community/pull/251#discussion_r1547438797), "If a feature is not able to transition to the next stage in the defined period, it should be removed automatically."

I think that it might take a while until these important features are evaluated, and they might get removed, if I understand correctly.

Reply:
The feature owner needs to consider this timeline. There is a balance that needs to be kept between adding features, keeping them in the queue for examination and finally graduating.
The motivation is describing that exact problem.

If the owner cannot prove and convince the maintainers the feature is ready for GA, it means it needs to drop. Kubernetes had the same problem of features remaining in Beta with no limit, which causes a maintenance mess (one one hand you do not commit to your users with a final contract and on the other hand keep the code in and cause all other to consider it in any integration).

Reply:
> * How can the feature owner or someone else prove whether the feature is being used or not?

This is out of scope.
But plain reasoning should work fine.

> * What's a reasonable amount of time for a feature to stay in testing phases like Alpha or Beta?

This proposal defines it in a table below and exceptions are possible with enough maintainer votes.

> * Should the time it takes to move from one phase to another be the same as the time given before removing unused features?

Again, it is defined in this proposal.

I do not understand why you comment about this at the motivation section.

Reply:
> This proposal defines it in a table below and exceptions are possible with enough maintainer votes.

Thanks, that makes sense. I missed it.

> But plain reasoning should work fine.

Could you provide a concrete example to help clarify?